Expecta/NSArray+Difference.m

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2020 Lightricks. All rights reserved.
+// Created by Maxim Grabarnik.
+
+#import "NSArray+Difference.h"
+
+#import <assert.h>
+
+#import "ExpectaSupport.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+API_AVAILABLE(ios(13.0))
+typedef NSOrderedCollectionChange<NSObject *> Change;
+
+API_AVAILABLE(ios(13.0))
+@interface ChangePair : NSObject
+
+- (instancetype)init NS_UNAVAILABLE;
+
+- (instancetype)initWithInsertion:(Change *)insertion removal:(Change *)removal;
+
+@property (nonatomic, readonly) Change *insertion;
+
+@property (nonatomic, readonly) Change *removal;
+
+@end
+
+@implementation ChangePair
+
+- (instancetype)initWithInsertion:(Change *)insertion removal:(Change *)removal {
+  if (self = [super init]) {
+    _insertion = insertion;
+    _removal = removal;
+  }
+  return self;
+}
+
+@end
+
+@implementation NSArray (Difference)
+
+- (NSArray<NSString *> *)differenceFrom:(id)object
+    API_AVAILABLE(ios(13.0)) {
+  assert([object isKindOfClass:NSArray.class]);
+  NSArray *expected = self;
+  NSArray *actual = (NSArray *)object;
+
+  NSOrderedCollectionDifference<NSObject *> *diff = [expected differenceFromArray:actual withOptions:NSOrderedCollectionDifferenceCalculationInferMoves];
+
+  NSMutableArray *mismatches = [NSMutableArray array];
+  NSMutableArray *reorders = [NSMutableArray array];
+  NSMutableSet *usedInsertions = [NSMutableSet set];
+  NSMutableSet *usedRemovals = [NSMutableSet set];
+  for (Change *insertion in diff.insertions) {
+    for (Change *removal in diff.removals) {
+      if (insertion.index == removal.index) {
+        [mismatches addObject:[[ChangePair alloc] initWithInsertion:insertion removal:removal]];
+        [usedInsertions addObject:@(insertion.index)];
+        [usedRemovals addObject:@(removal.index)];
+      }
+      if (insertion.associatedIndex == removal.index) {
+        assert(removal.associatedIndex == insertion.index);
+        [reorders addObject:[[ChangePair alloc] initWithInsertion:insertion removal:removal]];
+        [usedInsertions addObject:@(insertion.index)];
+        [usedRemovals addObject:@(removal.index)];
+      }
+    }
+  }
+
+  NSMutableArray *differences = [NSMutableArray array];
+  for (ChangePair *mismatch in mismatches) {
+    [differences addObject:[NSString stringWithFormat:@"Mismatch at index (%lu)",
+                            mismatch.insertion.index]];
+    if ([mismatch.insertion.object respondsToSelector:@selector(differenceFrom:)]) {
+      NSArray<NSString *> *diffLines = [(id)mismatch.insertion.object differenceFrom:mismatch.removal.object];
+      for (NSString *line in diffLines) {
+        [differences addObject:[NSString stringWithFormat:@"\t%@", line]];
+      }
```

Comment:
@barakwei Looking at it once again I think here we should prepend the inner diff with class name, else it would be very unclear.
We could also wrap `differenceFrom:` call in try/catch and fall back to default, but assertion with type mismatch message is probably better.
WDYT?

Reply:
I didn't understand why try/catch will help us.
in 99% of the cases the class names are the same in array, it may create noise.

Reply:
It's not about the array itself but values it holds. Let's say you have a `BlaModel` type with 3 fields named `field1-3` that also implements `differenceFrom`.
If I understand the code I wrote computing difference between [BlaModel('a','b','c'), @"stam"] and [BlaModel('a','D','c'), @"stam"] should be:

```
Mismatch at index (2)
	field2:
		Expected: D
		Actual: b
```

All the "field2..." is the deference between 2 `BlaModel` objects. However, the name of the type is not shown.

Another case is when `BlaModel` is compared to some `Foo`. Then there will be an assertion with message saying that "expected BlaModel type but got Foo", which is a good outcome.

Did I make it clearer or worse ?